core/kernel/tee_ta_manager.c

```diff
@@ -638,20 +670,31 @@ TEE_Result tee_ta_open_session(TEE_ErrorOrigin *err,
 
 TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
 				 struct tee_ta_session *sess,
+				 struct tee_ta_session_head *open_sessions,
 				 const TEE_Identity *clnt_id,
 				 uint32_t cancel_req_to, uint32_t cmd,
 				 struct tee_ta_param *param)
 {
 	TEE_Result res;
+	struct tee_ta_session *s;
 
 	if (check_client(sess, clnt_id) != TEE_SUCCESS)
 		return TEE_ERROR_BAD_PARAMETERS; /* intentional generic error */
 
 	if (!check_params(sess, param))
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	if (sess->ctx->panicked) {
+	if (!sess->ctx) {
 		DMSG("Panicked !");
+		*err = TEE_ORIGIN_TEE;
+		return TEE_ERROR_TARGET_DEAD;
+	} else if (sess->ctx->panicked) {
+		TAILQ_FOREACH(s, open_sessions, link)
+			if (s->ctx == sess->ctx && s != sess)
+				s->ctx = NULL;
+		tee_ta_context_destroy(sess->ctx);
+		sess->ctx = NULL;
```

Comment:
Can you make a function out of this sequence, with a nice explicit function label. It's used twice.

Add missing `mutex_lock(&tee_ta_mutex);` / `mutex_unlock(&tee_ta_mutex);`.

Reply:
Lately updated `tee_ta_destroy_context` has this locking/unlocking mechanism in place.